tests/test_gradient.py

```diff
@@ -139,12 +139,15 @@ def test_gradient_equivalence(self, shape, kernel, space_order, preset, nbl, dty
         assert(np.allclose(grad_u.data, grad_v.data, rtol=tolerance, atol=tolerance))
         assert(np.allclose(grad_u.data, grad_uv.data, rtol=tolerance, atol=tolerance))
 
-    @pytest.mark.parametrize('dtype', [np.float32, np.float64])
-    @pytest.mark.parametrize('space_order', [4])
-    @pytest.mark.parametrize('kernel', ['OT2'])
-    @pytest.mark.parametrize('shape', [(70, 80)])
-    @pytest.mark.parametrize('checkpointing', [True, False])
-    def test_gradientFWI(self, dtype, shape, kernel, space_order, checkpointing):
+    @pytest.mark.parametrize('dtype, space_order, kernel, shape, ckp, setup_func', [
+        (np.float32, 4, 'OT2', (70, 80), True, setup),
+        (np.float32, 4, 'OT2', (70, 80), False, setup),
+        (np.float64, 4, 'OT2', (70, 80), True, setup),
+        (np.float64, 4, 'OT2', (70, 80), False, setup),
+        (np.float64, 4, 'centered', (70, 80), True, tti_setup),
+        (np.float64, 4, 'centered', (70, 80), False, tti_setup),
```

Comment:
no float32?

Reply:
I forgot to add `np.float32`. I just tried it and the test fail (`AssertionError`) for both `Checkpointing=True and False` even for the first order error 
```
Operator `initdamp` run in 0.01 s
Operator `pad_vp` run in 0.01 s
Operator `pad_epsilon` run in 0.01 s
Operator `pad_delta` run in 0.01 s
Operator `pad_theta` run in 0.01 s
Operator `smoother` run in 0.01 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.03 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.03 s
1st order error, Phi(m0+dm)-Phi(m0): [ 0.87556505 -5.82210223]
2nd order error, Phi(m0+dm)-Phi(m0) - <J(m0)^T \delta d, dm>: [ 0.87554379 -5.82207074]
FOperator `initdamp` run in 0.01 s
Operator `pad_vp` run in 0.01 s
Operator `pad_epsilon` run in 0.01 s
Operator `pad_delta` run in 0.01 s
Operator `pad_theta` run in 0.01 s
Operator `smoother` run in 0.01 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.02 s
Operator `GradientTTI` run in 0.03 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.02 s
Operator `ForwardTTI` run in 0.02 s
1st order error, Phi(m0+dm)-Phi(m0): [ 0.87556505 -5.82210223]
2nd order error, Phi(m0+dm)-Phi(m0) - <J(m0)^T \delta d, dm>: [ 0.87554379 -5.82207074]
```
I don't really know what happened as with `np.float64` the test run without problems

```
Operator `initdamp` run in 0.01 s
Operator `pad_vp` run in 0.01 s
Operator `pad_epsilon` run in 0.01 s
Operator `pad_delta` run in 0.01 s
Operator `pad_theta` run in 0.01 s
Operator `smoother` run in 0.01 s
Operator `ForwardTTI` run in 0.03 s
Operator `ForwardTTI` run in 0.04 s
Operator `ForwardTTI` run in 0.12 s
Operator `ForwardTTI` run in 0.07 s
Operator `ForwardTTI` run in 0.07 s
Operator `ForwardTTI` run in 0.07 s
Operator `ForwardTTI` run in 0.06 s
Operator `ForwardTTI` run in 0.06 s
Operator `ForwardTTI` run in 0.07 s
1st order error, Phi(m0+dm)-Phi(m0): [  0.92561436 -10.858962  ]
2nd order error, Phi(m0+dm)-Phi(m0) - <J(m0)^T \delta d, dm>: [  1.97902976 -10.94126684]
.Operator `initdamp` run in 0.01 s
Operator `pad_vp` run in 0.01 s
Operator `pad_epsilon` run in 0.01 s
Operator `pad_delta` run in 0.01 s
Operator `pad_theta` run in 0.01 s
Operator `smoother` run in 0.01 s
Operator `ForwardTTI` run in 0.03 s
Operator `ForwardTTI` run in 0.05 s
Operator `GradientTTI` run in 0.05 s
Operator `ForwardTTI` run in 0.05 s
Operator `ForwardTTI` run in 0.05 s
Operator `ForwardTTI` run in 0.06 s
Operator `ForwardTTI` run in 0.06 s
Operator `ForwardTTI` run in 0.07 s
Operator `ForwardTTI` run in 0.07 s
Operator `ForwardTTI` run in 0.07 s
1st order error, Phi(m0+dm)-Phi(m0): [  0.92561436 -10.858962  ]
2nd order error, Phi(m0+dm)-Phi(m0) - <J(m0)^T \delta d, dm>: [  1.97902976 -10.94126684]
```
I'll have to look it up and that might take a while

Reply:
I added a new commit with the test for `np.float32 `and several tests from CI failed. I review the code in `operators.py` and `wavesolver.py `and I couldn't find an error. Then, I tried without the optimization flags (`-O3 and -ffast-math`) but the test still fails. @mloubout if you have any suggestions or comments about this, please let me know.

Reply:
There is definitely something off, I would understand if things were a bit off but that's way off. Can you plot the convergence curves and such to see how it behaves and maybe make sure model/data looks ok.

Reply:
I discovered what the problem was. The `tti_setup` function uses a default value of `tn=250.0`, so only the direct wave is registered, doubling the time solves the problem, thx.

Reply:
Ah, now I see why the bump to 500... OK. Instead of changing it from `tti_setup` and `run`, can you pass the value from the tests only? also, is there a smaller value such that the forward wave is recorded, such as 300 or 350? or a smaller grid? we need to bring down the turnaround time of these tests

Reply:
I've reduced the grid dimensions as well as the recording time to 400.